### DISCUSSION TITLE
Allow using 'typeof' keyword with definitely typed variable/property as a parameter

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#2847] [Language Ideas]
<p>Let's assume, you have some POCO class with few properties:</p><pre>public class FooBar
{
  public int Foo { get; set; }
  public string Bar { get; set; }
}</pre><p>And you need to generate and store some metadata for this class.</p><p>You can write:</p><pre>var MemberName = nameof(FooBar.Foo);
</pre><p>But you can't use:</p><pre>var MemberType = typeof(FooBar.Foo);</pre><p>Although compiler has all information about Foo type (unless Foo is dynamic).</p><p>You need to write instead:</p><pre>var MemberType = typeof(int);
</pre><p>And this will be definitely broken during refactoring, if you will change the type of Foo in POCO.</p><p>Moreover, in this situation, 'nameof' and 'typeof' implementation seems to have some inconsistency, as for me.</p><p>So my suggestion is to implemement and allow using of</p><pre>string variable;
var type = typeof(variable); // type = System.String</pre><p>At least in case when variable type can be obtained during compilation.</p>

_This issue has been moved from https://developercommunity.visualstudio.com/content/idea/389549/allow-using-typeof-keyword-with-definitely-typed-v.html
VSTS ticketId: 733912_
_These are the original issue comments:_

Jane Wu [MSFT] on 11/23/2018, 00:14 AM (12 days ago): <p>Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.&nbsp;You will hear from us in about a week on our next steps.</p>
Vladyslav Prokopenko on 11/23/2018, 02:27 AM (12 days ago): <p>Looks reasonable and natural.<br>As alternative solution I can propose to use <a target='_blank' href="https://docs.microsoft.com/en-us/dotnet/api/system.object.gettype?view=netframework-4.7.2" target="_blank">GetType</a>.</p>

Anton Zhaparov on 11/23/2018, 02:58 AM (12 days ago): <p><em>GetType </em>is an instance method, so it works only with class instance, not with type definition itself. </p><p>Thus to utilize it, I will need to create a dummy instance of the POCO class... Which is possible, but complicates the code, while more straightforward and elegant solution can be implemented, I believe.</p>

Vladyslav Prokopenko on 11/23/2018, 03:05 AM (12 days ago): <p>Sure, <em>GetType</em> is not ideal solution, but for POCO's it'll work in most cases, as the class is trivially-constructible.<br>For other cases only reflection will help you, unfortunately.</p>

**Comment by svick** [comment]
This is largely a duplicate of https://github.com/dotnet/csharplang/issues/988.

**Comment by GreemDev** [comment]
`typeof(variable)` Why use this when any given variable has `.GetType()`? I mean, it looks subjectively better, but it'd just be duplicate functionality.

Edit: debunked below, thanks!

**Comment by yaakov-h** [comment]
`GetType()` will always return the most derived type, which might not be what you want.
`GetType()` doesn't work on `null`, because `null` is not an object and has no type.
`GetType()` will box a value type.
`GetType()` needs to be virtually dispatched and is thus slightly slower.

Need I continue?

**Comment by YairHalberstadt** [comment]
> GetType() will box a value type.

This is not true as of .Net Core 2.2

Not that it changes the thrust of your argument. `GetType` returns the actual type of an object. `typeof(variable)` would return the static type of a named storage location. They are two entirely different things.

**Comment by CyrusNajmabadi** [comment]
> GetType() will box a value type.

That shouldn't be the case.  Indeed, the JIT will simply replace the entire operation with the actual type in-situ.  There will no actual cost beyond that when the code executes.

**Comment by yaakov-h** [comment]
Not on 32-bit .NET Framework. 😢 

But yes, 64-bit .NET Framework and .NET Core are smarter about it.

**Comment by GreemDev** [comment]
> `GetType()` will always return the most derived type, which might not be what you want.
> `GetType()` doesn't work on `null`, because `null` is not an object and has no type.
> `GetType()` will box a value type.
> `GetType()` needs to be virtually dispatched and is thus slightly slower.
> 
> Need I continue?

Ah, I didn't know this stuff. Thanks!

**Comment by RamType0** [comment]
```C#
using VariableType = Foo.Bar;
var type = typeof(VariableType);
```
